src/setup/node.rs

```diff
@@ -0,0 +1,167 @@
+//! High level APIs and types for node setup and teardown.
+
+use std::{
+    fs, io,
+    path::{Path, PathBuf},
+    process::{Child, Command, Stdio},
+};
+
+use anyhow::Result;
+use fs_extra::dir;
+
+use super::constants::{ALGORAND_SETUP_DIR, NODE_DIR, PRIVATE_NETWORK_DIR};
+use crate::setup::{
+    config::{NodeConfig, NodeMetaData},
+    get_algorand_work_path,
+};
+
+pub enum ChildExitCode {
+    Success,
+    ErrorCode(Option<i32>),
+}
+
+pub struct NodeBuilder {
+    /// Node's startup configuration.
+    conf: NodeConfig,
+    /// Node's process metadata read from Ziggurat configuration files.
+    meta: NodeMetaData,
+}
+
+impl NodeBuilder {
+    /// Creates a new [NodeBuilder].
+    pub fn new() -> anyhow::Result<Self> {
+        let setup_path = get_algorand_work_path()?.join(ALGORAND_SETUP_DIR);
+
+        let conf = NodeConfig::default();
+        let meta = NodeMetaData::new(setup_path)?;
+
+        Ok(Self { conf, meta })
+    }
+
+    /// Creates a [Node] according to configuration and starts its process.
+    pub async fn start(&mut self, target: &Path) -> Result<Node> {
+        if !target.exists() {
+            fs::create_dir_all(&target)?;
+        }
+        let setup_path = get_algorand_work_path()?.join(ALGORAND_SETUP_DIR);
+
+        // Currently we can start only the first node
+        let source = get_node_path(0)?;
+
+        let mut copy_options = dir::CopyOptions::new();
+        copy_options.content_only = true;
+        copy_options.overwrite = true;
+        dir::copy(&source, &target, &copy_options)?;
+
+        // TODO(Rqnsom) configure the node
+
+        self.meta.start_args.push("-d".into());
+        self.meta.start_args.push(target.into());
```

Comment:
needs a comment

Reply:
also, I think we should be modifying these in the target node; the `&mut self` that this enforces here is both surprising and misleading, as we could spawn multiple nodes with this builder; alternatively, we could consume the builder when starting the node, in which case we could take `mut self`.

Reply:
> as we could spawn multiple identical nodes with this builder

Tbh, I don't like starting a node from the `NodeBuilder`. We could make it so that the node builder spawns a node instance. Then, we start it from the Node via the `start` call.

Suggestion:
```
let node = nodeBuilder->spawn();
node.start()?;
```

With such a solution, we could move this part into the Node then.

Reply:
I think the usual convention is to call `build` on a builder in order to create the target object; other than that, I'm ok with the node then being started individually.